torch/csrc/jit/python/init.cpp

```diff
@@ -885,13 +886,125 @@ void initJITBindings(PyObject* module) {
             docstring << "  " << op->schema() << "\n";
           }
 
-          return py::cpp_function(
-              [operations](py::args args, py::kwargs kwargs) {
+          // To support __torch_function__ override mechanism, `func` needs to
+          // provide a pointer, which points to the original method, i.e.,
+          // itself. py::cpp_function doesn't support specifying the function
+          // signature without real implementation, so we we use a trick here.
+          // First, we create a function holder pointing to an empty
+          // cpp_function. Then we use a reference as the alias of the empty
+          // function. Then we define the real returned function, and update
+          // the value of the function pointed by the holder using the alias.
+          // With this trick, we are able to call self inside the returned
+          // function.
+          auto func_holder = std::make_unique<py::cpp_function>();
+          py::cpp_function& func_ref = *func_holder;
+          auto func = py::cpp_function(
+              [operations, op_name, self_ptr = std::move(func_holder)](
+                  py::args args, py::kwargs kwargs) {
+                std::shared_ptr<Operator> found_op = nullptr;
+                if (operations.size() == 1) {
+                  found_op = operations.at(0);
+                } else {
+                  for (const auto& op : operations) {
```

Comment:
this matching is still not ideal - what if there are two overrides:
```
foo(x : Tensor)
foo(x : List[Tensor])
```

Since you don't verify the matching here, it's going to pick the first overload and miss iterating over the list, right?

I don't really get why we need to do schema matching here at all, for a baseline implementation we could just iterate blindly over the entire list of arguments (as you do anyway) and if it's Python's tuple or list - then recurse